### ISSUE TITLE
NaN produced in `log(2 * pi * s$sigma2)`

### ISSUE THREAD

**Issue by gaow** [stephenslab/susieR#7]
I'm just reporting this issue for now. I'll investigate it next. 

## MWE

[susie_issue6.tar.gz](https://github.com/stephenslab/susieR/files/1992536/susie_issue6.tar.gz)

## Code

```r
attach(readRDS('susie_issue6.rds'))
susieR::susie(X,Y)
```

## Outcome

```
Warning message in log(2 * pi * s$sigma2):
“NaNs produced”
Error in if ((elbo[i + 1] - elbo[i]) < tol) break: missing value where TRUE/FALSE needed
Traceback:

1. susieR::susie(X, Y)
```

**Comment by gaow** [comment]
With:

```diff
index 0420271..37e1a09 100644
--- a/R/estimate_residual_variance.R
+++ b/R/estimate_residual_variance.R
@@ -4,6 +4,7 @@
 #' @param s a susie fit
 estimate_residual_variance = function(X,Y, s){
   n = nrow(X)
+  print(get_ER2(X,Y,s))
   return( (1/n)* get_ER2(X,Y,s) )
 }
```

I get on my screen

```
[1] -1.472293e+12
```

So that leads to a very negative `sigma2` estimate.

**Comment by stephens999** [comment]
i get no problem running this example with susieR_0.1.6

**Comment by gaow** [comment]
Hmm this is weird, because I believe I'm on the same version. Here I also attach the session info:

```r
> attach(readRDS('susie_issue6.rds'))
> susieR::susie(X,Y)
Error in if ((elbo[i + 1] - elbo[i]) < tol) break : 
  missing value where TRUE/FALSE needed
In addition: Warning message:
In log(2 * pi * s$sigma2) : NaNs produced



> print(sessionInfo())
R version 3.4.4 (2018-03-15)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: BunsenLabs GNU/Linux 8.7 (Hydrogen)

Matrix products: default
BLAS/LAPACK: /opt/intel/compilers_and_libraries_2018.2.199/linux/mkl/lib/intel64_lin/libmkl_rt.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

loaded via a namespace (and not attached):
[1] compiler_3.4.4 susieR_0.1.6  
> 
```

**Comment by gaow** [comment]
Hmm this is really weird, but seems linking BLAS to Intel MKL is the culprit. I suspected that, and to test it out I built a [docker image for susieR](https://hub.docker.com/r/gaow/susie/) that uses standard Debian stock R software, and ran this command:

```bash
docker run --rm gaow/susie Rscript -e "attach(readRDS('/opt/susie_issue6.rds')); res = susieR::susie(X,Y); print(names(res)); print(sessionInfo())"
```

and I got no issues!

```r
 [1] "alpha"                  "mu"                     "mu2"                   
 [4] "Xr"                     "sigma2"                 "sa2"                   
 [7] "KL"                     "elbo"                   "intercept"             
[10] "fitted"                 "X_column_scale_factors"
R version 3.3.3 (2017-03-06)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Debian GNU/Linux 9 (stretch)

locale:
[1] C

attached base packages:
[1] stats     graphics  grDevices utils     datasets  base     

loaded via a namespace (and not attached):
[1] susieR_0.1.6
```

(see the R version there)

This is a rare case but I guess it just happens ... I'll try change my R version and see how it works.

**Comment by gaow** [comment]
Okay, confirmed -- after re-linking my BLAS from Intel MKL to ATLAS the problem goes away! See below:

```r
> attach(readRDS('susie_issue6.rds'))
> res = susieR::susie(X,Y)
> sessionInfo()
R version 3.4.4 (2018-03-15)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: BunsenLabs GNU/Linux 8.7 (Hydrogen)

Matrix products: default
BLAS: /usr/lib/atlas-base/atlas/libblas.so.3.0
LAPACK: /usr/lib/atlas-base/atlas/liblapack.so.3.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

loaded via a namespace (and not attached):
[1] compiler_3.4.4 susieR_0.1.6 
```

So I wasted a good chunk of the day chasing this wild bug ... there certainly is a lesson learned, a weird one.

**Comment by stephens999** [comment]
ok, but might still be worth to understand the bug? If it is our bug?

**Comment by gaow** [comment]
> ok, but might still be worth to understand the bug? If it is our bug?

Definitely. That is why I left this ticket open. I actually tried some more last night. If the Intel MKL is pre-compiled with R then there is no issue. So that'd be a bit reassuring. It seems it only goes wrong when R is not compiled or statically linked with MKL but only dynamically linked (or re-linked).

Shall we still keep this ticket open so that we can investigate down the line, or run into similar issues?

**Comment by stephens999** [comment]
Sounds good

**Comment by KaiqianZhang** [comment]
I have a similar error when trying to fit susie with a L0 initialization. 

`test.s.init = readRDS('test_s_init.rds')
  X.fit.susie = susie(X,y,s_init=test.s.init)
`
and I got the following error:

> Error in if ((elbo[i + 1] - elbo[i]) < tol) break : 
  missing value where TRUE/FALSE needed
[test_s_init.rds.zip](https://github.com/stephenslab/susieR/files/2177983/test_s_init.rds.zip)

**Comment by KaiqianZhang** [comment]
Here is my code for X, y and test.s.init:

```
devtools::install_github("stephenslab/susieR")
devtools::install_github("hazimehh/L0Learn")
library(susieR)
library(L0Learn)
```

```
set.seed(1)
n = 1000
p = 1000
beta = rep(0,p)
beta[1] = 1
beta[2] = 1
beta[3] = 1
beta[4] = 1
X = matrix(rnorm(n*p),nrow=n,ncol=p)
y = X %*% beta + rnorm(n)
res =susie(X,y,L=10)
```

```
susie_L0_init = function(effect_index, effect_beta, p, prior_variance, residual_variance){
  L=length(effect_index)
  alpha=matrix(0,nrow=L,ncol=p)
  mu=matrix(0,nrow=L,ncol=p)
  for(i in 1:L){
    alpha[i, effect_index[i]]=1
    mu[i, effect_index[i]]=effect_beta[i]
  }
  mu2 = mu*mu
  sigma2 = residual_variance
  sa2 = prior_variance
  KL = rep(NA,L)
  Xr = colSums(alpha*mu)
  s = list(alpha=alpha, mu=mu, mu2=mu2, sigma2=sigma2, sa2=sa2, KL=KL, Xr=Xr)
  class(s) <- "susie"
  return(s)
}
```

```
test.fit.L0 = L0Learn.cvfit(X, y, penalty = "L0")
```

```
lambdaIndex = which.min(test.fit.L0$cvMeans[[1]]) 
test.coef.L0 = coef(test.fit.L0$fit, lambda = test.fit.L0$fit$lambda[[1]][lambdaIndex])
test.effect.beta = test.coef.L0[which(test.coef.L0!=0)][-1]
test.effect.index = (which(test.coef.L0!=0)-1)[-1] 
```

```
test.s.init = susie_L0_init(test.effect.index, test.effect.beta, p=1000, prior_variance=0.2, residual_variance=as.numeric(var(y)))
```
You can simply run rds for test.s.init if you do not want to load L0Learn package. 

It seems like that `susie_get_objective(X,Y,s)` this function returns NA.

**Comment by pcarbo** [comment]
@KaiqianZhang I don't understand the issue---where does the error occur? Where are you calling `susie_get_objective`? I ran `res = susie(X,y,L = 10)` and did not get an error.

**Comment by KaiqianZhang** [comment]
@pcarbo `res = susie(X,y,L = 10)` would not produce error. But if you initialize it with L0 fit, e.g. `res = susie(X,y,L = 10, s_init=test.s.init)`, you will get the error.

`susie_get_objective` is called inside susie. I printed the output of this function and found it gave me NA.

**Comment by pcarbo** [comment]
@KaiqianZhang I got a different error, but I can at least say it is not working for me. It seems that initializing SuSiE requires some care; perhaps @gaow can provide some guidance.

**Comment by gaow** [comment]
@pcarbo Kaiqian and I are mostly working together (we are door to door). This is the 2nd top priority on our to-do list I believe and we'll revisit offline soon before reporting back some more here.

**Comment by stephens999** [comment]
I tried it and got the same error (I had to change to coef output using
as.vector).
 But I am not convinced it was necessarily "caused" by initialization.
 L was quite large here, so maybe that is part of the cause.

@Kaiqian Zhang <kaiqianz@uchicago.edu> can you try with an L0
initialization with smaller L (say 10)?

**Comment by stephens999** [comment]
@Kaiqian Zhang <kaiqianz@uchicago.edu> also can you make everything
reproducible by setting a seed at start of code? L0learn may not be
deterministic, i'm not sure...

**Comment by KaiqianZhang** [comment]
@stephens999 I set.seed(1) in the beginning when I was simulating the data. I do not know where else I should set a seed? Before running L0Learn fit?

**Comment by KaiqianZhang** [comment]
@stephens999 Since L0 initialization provides 115 variables, I selected the first ten [1:10](i.e set L=10 as you suggested). But I still get the same error. 

Here is my code. Note that test.effect.beta and  test.effect.index are what I got before from L0Learn fit.
```
test.effect.beta.short = test.effect.beta[1:10]
test.effect.index.short = test.effect.index[1:10]
test.s.init = susie_L0_init(test.effect.index.short, test.effect.beta.short, p=1000, prior_variance=0.2, residual_variance=as.numeric(var(y)))
X.fit.susie = susie(X,y,s_init=test.s.init)
```
Same error:

> Error in if ((elbo[i + 1] - elbo[i]) < tol) break : 
  missing value where TRUE/FALSE needed

**Comment by pcarbo** [comment]
@KaiqianZhang How you set the seed should be sufficient. I think Matthew missed the `set.seed` call at the top of your script.

**Comment by gaow** [comment]
Turns out to be a minor problem in the initialization code, not related to susieR (@KaiqianZhang  :P). So unfortunately (or fortunately) we still cannot reproduce the susieR behavior I initially ran into (see title of this ticket). I'll close the ticket for now. We can reopen if we can reproduce the ticket with a legit example in the future.

**Comment by zouyuxin** [comment]
I have the same error. The data set and the R script is: 
[ELBO NaN.zip](https://github.com/stephenslab/susieR/files/2858942/ELBO.NaN.zip)

It works on my MacBook, but fails at RCC. On RCC, the elbo becomes NaN (the sigma2 becomes 0) after several iterations.

**Comment by zouyuxin** [comment]
The data is generated under the null (no signals) using DSC with pre-randomly-generated X. The weired thing is that the simualted Y is same as the first column of X! It's possible that X and Y are generated under the same seed.

**Comment by gaow** [comment]
@zouyuxin the seed for the first DSC replicate is 1. Is that what you used to generate your `X`?

On my Linux desktop I can reproduce this issue:

```
$ Rscript ELBO_NaN.R 
Error in if ((elbo[i + 1] - elbo[i]) < tol) { : 
  missing value where TRUE/FALSE needed
Calls: susie
Execution halted
```

my R: 

```r
> library(susieR)
> sessionInfo()
R version 3.5.1 (2018-07-02)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: BunsenLabs GNU/Linux 8.9 (Hydrogen)

Matrix products: default
BLAS: /usr/lib/openblas-base/libblas.so.3
LAPACK: /usr/lib/libopenblasp-r0.2.19.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] susieR_0.6.4.0438

loaded via a namespace (and not attached):
[1] compiler_3.5.1  Matrix_1.2-14   grid_3.5.1      lattice_0.20-35
```

**Comment by zouyuxin** [comment]
>the seed for the first DSC replicate is 1. Is that what you used to generate your `X`?

yes~

**Comment by pcarbo** [comment]
I can confirm this issue in R 3.5.1 on midway2 (using susieR 0.9.55).

**Comment by pcarbo** [comment]
I've addressed this issue in commit 15a0293. I'm leaving this issue since @stephens999 may have some additional improvements he would like to incorporate relevant to this.
